python/bifrost/blocks/psrdada.py

```diff
@@ -129,7 +147,174 @@ def on_data(self, reader, ospans):
         nframe = nbyte // ospan.frame_nbyte
         return [nframe]
 
-def read_psrdada_buffer(buffer_key, header_callback, gulp_nframe, space=None,
+def _keyval_to_dadastr(key, val):
+     """ Convert key: value pair into a DADA string """
+     return "{key:20s} {val}\n".format(key=key.upper(), val=val)
+
+
+def _extract_tensor_scale(key, tensor):
+    try:
+        idx = tensor['labels'].index(key)
+        return tensor['scales'][idx]
+    except ValueError as e:
+        return [0, 0]
+
+
+def _extract_tensor_shape(key, tensor):
+    try:
+        idx = tensor['labels'].index(key)
+        return tensor['shape'][idx]
+    except ValueError as e:
+        return 1
+
+     
+def generate_dada_header(hdr_dict, hdrlen=4096):
+    """ Generate DADA header from header dict 
+    
+    Args:
+        hdr_dict (dict): Header dictionary of key, value pairs
+        hdrlen (int): Size of header, default 4096
+    
+    Returns:
+        s (str): DADA header string with padding to hdrlen
+    """
+    s = "HDR_VERSION         1.0\n"
+    s+= "HDR_SIZE            %i\n" % hdrlen
+    keys_to_skip = ('HDR_VERSION', 'HDR_SIZE')
+
+    # update parameters from bifrost tensor
+    if '_tensor' in hdr_dict.keys():
+        print(hdr_dict['_tensor'])
+        dtype = hdr_dict['_tensor']['dtype']
+        dtype_vals = {
+            'cf32': { 'NBIT': '32', 'NDIM': '2' },
+            'f32': { 'NBIT': '32', 'NDIM': '1' },
+            'ci8': { 'NBIT': '8', 'NDIM': '2' },
+            'i8': { 'NBIT': '8', 'NDIM': '1' } }
```

Comment:
Are additional data types expected/allowed?

Reply:
There isn't really a strict definition AFAIK. There is no header keyword for float/integer, but I think it's reasonable to add cf64/f64. Would need to decide between i16/ci16 and f16/cf16 though

Reply:
i16/ci16 and f64/cf64 sound good to me.  I'm not sure that bifrost works with f16/cf16, at least on the cuda side.  I should look into that.